File: amicable_pairs.rb
#!/usr/bin/env ruby
# http://rosettacode.org/wiki/Amicable_pairs

def factors(n)
  n_factors =[]
  (1..Math.sqrt(n)).select { |f| n % f == 0 }.each { |v| n_factors << n/v; n_factors << v }
  n_factors.uniq.sort - [n]
end

def sum(array,index=0)
  index == array.length ? 0 : array[index] + sum(array,index+1)
end

def find_amicable_pairs(limit)
  pairs = []; amicables = {}
  (1...limit).each { |n| amicables[n] = sum(factors(n)) }
  amicables.each { |k,v| pairs << [k,v].sort if amicables[v] == k && v != k }
  pairs.uniq
end

p find_amicable_pairs(20_000)

File: amicable_pairs.swift
#!/usr/bin/env swift
// http://rosettacode.org/wiki/Amicable_pairs

func factors(n: Int) -> [Int] {
  var factorsOfN: [Int] = []
  for i in 1..<n {
    if n % i == 0 { factorsOfN.append(i) }
  }
  return factorsOfN
}

func findAmicablePairs(limit: Int) -> [Int:Int] {
  var pairs: [Int:Int] = [:]
  var amicables:[Int:Int] = [:]
  for i in 1..<limit {
    amicables[i] = factors(i).reduce(0) { $0! + $1 }
  }
  for (k,v) in amicables {
    if amicables[v] == k && v != k {
      let temp = [k,v].sort { $0 < $1 }
      pairs[temp[0]] = temp[1]
    }
  }
  return pairs
}

print(findAmicablePairs(20_000))
